Perf capture bug fixes

kbrock opened the pull request:
Part of https://github.com/ManageIQ/manageiq/issues/20071

depends:
- [x] https://github.com/ManageIQ/manageiq/pull/22184

Fix
- if multiple vms are collected for the same timestamp, metrics for only one of vms was collected

Enhancement:
- Don't lookup collected objects as second time in perf_process
- lotz o' comments
- 
### ~~lookup zones~~ removed

### lookup objects

|       ms |       bytes |   objects |query |  qry ms |     rows |comments
|      ---:|         ---:|       ---:|  ---:|     ---:|      ---:| ---
|  2,787.3 | 19,375,911* | 1,587,039 |  610 | 1,026.8 |      288 |capture-before
|  2,600.0 | 18,166,833* | 1,557,973 |  588 |   948.0 |      266 |capture-after
|       7% |             |           |   4% |      8% |       8% | delta

kbrock commented on app/models/miq_region.rb:
```diff
@@ -99,6 +99,10 @@ def find_master_server
 
   cache_with_timeout(:my_region) { find_by(:region => my_region_number) }
 
+  cache_with_timeout(:my_zone_names) do
+    Zone.in_region(my_region_number).pluck(:name)
+  end
```
Alternatively, we can just use `my_region` as the cache and store in the object:

```ruby
  def zone_names do
    @zone_names ||= Zone.in_region(region_number).pluck(:name)
  end
```

Fryguy replied:
Yes, I'd rather not introduce a new cache when one already exists.

kbrock replied:
I did that - using `MiqRegion.zones` cache

kbrock replied:
moved this stuff to another PR

agrare commented on app/models/metric/ci_mixin/processing.rb:
```diff
@@ -105,6 +105,9 @@ def perf_process(interval_name, start_time, end_time, counters_data)
 
   def transform_resources!(counters_data)
+    # no work needed if keys are already ActiveRecord objects
+    return if counters_data.first.first.kind_of?(ActiveRecord::Base)
```
`counters_data` is a hash right and you're checking the first key?  Wonder if `counters_data.keys.first` would be more obvious as to what we're actually doing here

agrare replied:
Do we need to check `keys.all? { |k| k.kind_of?(ActiveRecord::Base) }` or is it not possible to have mixed types?

kbrock replied:
yea. `first.first` is more efficient than `keys.first` but I think readability is best for us, especially since we are limiting this to 250 objects (for vmware at least)

kbrock replied:
updated: I couldn't get this code the way I wanted it.
Decided to delete instead

kbrock commented on app/models/metric/ci_mixin/processing.rb:
```diff
@@ -98,24 +94,12 @@ def perf_process(interval_name, start_time, end_time, counters_data)
     end
     _log.info("#{log_header} Processing for #{log_specific_targets(resources)}, for range [#{start_time} - #{end_time}]...Complete - Timings: #{t.inspect}")
-
-    affected_timestamps
```
I had changed to:

```diff
- affected_timestamps
+ affected_timestamps.uniq
```

question: Is this was intentional.
answer: yes

new answer: `affected_timestamps` is not used, so dropping it

kbrock commented on app/models/metric/ci_mixin/processing.rb:
```diff
@@ -98,24 +94,12 @@ def perf_process(interval_name, start_time, end_time, counters_data)
 
   private
 
-  def transform_resources!(counters_data)
-    # Fetch ActiveRecord object by 1 query per Model
-    grouped_resource_refs = counters_data.keys.each_with_object({}) { |x, obj| (obj[x.first] ||= []) << x.second }
-    fetched_records = grouped_resource_refs.keys.each_with_object({}) do |x, obj|
-      x.constantize.where(:id => grouped_resource_refs[x]).each { |rec| obj[[x, rec.id]] = rec }
-    end
-    # Transforming [Class, id] that were sent via the counters_data into the ActiveRecord objects
-    counters_data.transform_keys! { |x| fetched_records[x] }
-  end
```
We were sending `[class.name, id]` over the wire to this method.
I'm guessing that was to support sending these values over kafka to have a single point of saving.

We can bring this back if we decide to separate fetching of metrics with saving them.

agrare replied:
It looks like per 7eebbeb7dcd it was `# Convert to format allowing to send multiple resources at once` not really sure how that helps with that though :shrug:

kbrock reviewed (commented):
wanted to capture comments from @agrare

kbrock commented:
update:
- dropped `affected_timestamps` since the return from `perf_process` is not used anywhere
- dropped `transform_resources!` since we are passing the resources and no transform needed
